File: LeetCodeSolutions/LeetCode_0285.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode':
        used_p = False
        stack = [(root, False)]
        while stack:
            node, used = stack.pop()
            if node == p:
                used_p = True
            if not used:
                if node.right:
                    stack.append((node.right, False))
                stack.append((node, True))
                if node.left:
                    stack.append((node.left, False))
            else:
                if used_p and node.val > p.val:
                    return node
        return None

File: LeetCodeSolutions/LeetCode_0285.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        stack<pair<TreeNode*, bool> > s;
        if(!root) return nullptr;
        s.push(make_pair(root, false));
        bool isThis = false;
        while(!s.empty()){
            auto tmp = s.top();
            s.pop();
            if(!tmp.second){
                tmp.second = true;
                if(tmp.first->right) s.push(make_pair(tmp.first->right, false));
                s.push(tmp);
                if(tmp.first->left) s.push(make_pair(tmp.first->left, false));
            }
            else{
                if(isThis) return tmp.first;
                if(tmp.first == p) isThis = true;
            }
        }
        return nullptr;
    }
};